blog-posts/cloudfront-cache-efficiency.md

```diff
@@ -0,0 +1,154 @@
+and decide to not cache the object. Which is not what we want.
+
+After deploying the CloudFront function, let’s check the browser console:
+
+![Requests are cached by the browser](images/cloudfront-cache-efficiency/2-improved.png)
+
+What do we see here:
+
+1. The first request hits Lambda, so it takes around 700ms.
+2. The requests within 10 seconds are cached. After that, one request is done to CloudFront. This request takes around 30ms
+   and is cached again for 10 seconds by the browser. CloudFront returns the full response.
+3. Only after 1 minute, the requests hits Lambda again and gets then cached again by CloudFront.
+
+This is the behavior we want! Nice!
+
+Next, let’s see how we can return `304` if we hit CloudFront and the response hasn't changed since the last request.
+
+## Return 304 if the object hasn’t changed in CloudFront
+
+CloudFront always returns `200` with the full response body. This results in a lot of unnecessary data send between the
+browser and CloudFront. To fix this, we need to set the `ETag` header in our Lambda function.
+
+The `ETag` is a hash for the response body. The browser will send this hash with alongside with all upcoming requests to
+the same resource. The server can then decide if the resource has changed and either return `200` with a new resource
+or `304` .
+
+Luckily, when the `ETag` header is set, CloudFront will handle `200` and `304` automatically. The `ETag` need to be
+created in our lambda function. This is how it could look like:
+
+```js
+const createEtag = (body: string) => {
+  const hash = createHash('md5').update(body).digest('hex');
+  return `W/"${hash}"`
+}
+```
+
+Depending on your usage, you may want to use different hashing algorithms that are more performat.
+
+Now let’s check again our browser console:
+
+![CloudFront even returns 304](images/cloudfront-cache-efficiency/3-solution.png)
+
+Now it’s perfect!
+
+1. The first request hits Lambda. The response gets cached by the browser for 10 seconds and 60 seconds by CloudFront.
+2. The next 3 requests are within 10 seconds, so they are not send over the network.
+3. After 10 seconds, the request get’s send to CloudFront. CloudFront return `304` with an empty response body.
+4. After 60 seconds the requests hits Lambda again.
+
+## Conclusion
+
+CloudFront is a very good CDN, but you should always test if your requests are cached correctly.
+
+- If you have a different `s-maxage` and `maxage` you need to overwrite the `Date` and `Age` header in CloudFront.
+- If you want to return `304` if the response hasn't changed since the last request, you need to set the `ETag` header in
+your origin.
+
+> Checkout the full CloudFormation template on Github: [https://github.com/feedm3/learning-caching-headers/blob/main/serverless.yml#L19-L88](https://github.com/feedm3/learning-caching-headers/blob/main/serverless.yml#L19-L88)
```

Comment:
I think in general it would be good to store those testing repos in the satellytes org. Not to take ownership but not to blur the border between private and business? Do you think that would be possible here? It's not a blocker, just an idea. We could also introduce a naming scheme or some other way to collect those "test repos" if possible